File: analysis/python/path_generation.py
#!/usr/bin/env python3

from __future__ import annotations

import numpy as np

__all__ = ["path_generation"]


def path_generation(
    x0: float,
    xf: float,
    tf: float,
    ta: float,
    h: float,
    full_quintic: bool = False,
) -> np.ndarray:
    """
    각 샘플마다 ``[위치, 속도, 가속도]`` 인 ``(N, 3)`` 배열을 반환한다.

    - 열 0: ``x(t)``, 열 1: ``dx/dt``, 열 2: ``d²x/dt²`` (5차 다항의 해석 미분).
    - ``path_generation(...)[:-1]`` 로 슬라이스하면 세 양이 **같은 행** 단위로 잘린다.
    - 구간 끝 시각 ``tf`` 샘플은 **포함**된다(``round(tf/h)+1`` 점).

    ``full_quintic=False`` (기본): MATLAB 트래페조이드 — 가속·등속·감속 3구간 5차 연결.
    ``full_quintic=True``: ``x0``→``xf`` 를 시간 ``tf`` 안에 **한 번의** 5차 다항(휴지–휴지).
    """
    if full_quintic:
        return quintic_path(x0, xf, 0.0, 0.0, 0.0, 0.0, tf, h)

    td = tf - ta
    vd = (xf - x0) / td
    xa = x0 + 0.5 * ta * vd
    xd = xf - 0.5 * ta * vd

    pv1 = quintic_path(x0, xa, 0.0, vd, 0.0, 0.0, ta, h)
    pv2 = quintic_path(xa, xd, vd, vd, 0.0, 0.0, td - ta, h)
    pv3 = quintic_path(xd, xf, vd, 0.0, 0.0, 0.0, tf - td, h)
    return np.concatenate((pv1, pv2, pv3), axis=0)


def quintic_path(
    pos0: float,
    posf: float,
    vel0: float,
    velf: float,
    acc0: float,
    accf: float,
    ts: float,
    h: float,
) -> np.ndarray:
    """한 구간: ``ts<=0`` 이면 ``(0, 3)``.

    샘플 시각은 ``0`` 부터 ``ts`` **끝점 포함**이며, 간격 수는 ``round(ts/h)`` 로 두어
    ``h`` 로 나누어떨어지는 구간(예: ``ts=0.5``, ``h=0.001`` → 501점)에서 RecurDyn과 같은
    시간 격자 길이와 맞추기 쉽다. (``np.arange(0, ts, h)`` 는 끝 ``ts`` 가 빠져 한 점 부족해짐.)
    """
    if ts <= 0:
        return np.zeros((0, 3), dtype=float)

    a0, a1, a2, a3, a4, a5 = quintic_coeffs(pos0, posf, vel0, velf, acc0, accf, ts)
    n_interval = max(1, int(round(ts / float(h))))
    t_samples = np.linspace(0.0, ts, n_interval + 1, dtype=float)
    pos_list: list[float] = []
    vel_list: list[float] = []
    acc_list: list[float] = []
    for t in t_samples:
        pos_list.append(a0 + a1 * t + a2 * t**2 + a3 * t**3 + a4 * t**4 + a5 * t**5)
        vel_list.append(a1 + 2 * a2 * t + 3 * a3 * t**2 + 4 * a4 * t**3 + 5 * a5 * t**4)
        acc_list.append(2 * a2 + 6 * a3 * t + 12 * a4 * t**2 + 20 * a5 * t**3)
    pos = np.asarray(pos_list, dtype=float)
    vel = np.asarray(vel_list, dtype=float)
    acc = np.asarray(acc_list, dtype=float)
    return np.column_stack((pos, vel, acc))


def quintic_coeffs(
    pos0: float,
    posf: float,
    vel0: float,
    velf: float,
    acc0: float,
    accf: float,
    ts: float,
) -> tuple[float, float, float, float, float, float]:
    """5차 경로 다항식 계수 ``a0..a5`` (위치 기준)."""
    a0 = pos0
    a1 = vel0
    a2 = acc0 / 2.0
    a3 = (20 * (posf - pos0) - (8 * velf + 12 * vel0) * ts - (3 * acc0 - accf) * ts**2) / (
        2 * ts**3
    )
    a4 = (30 * (pos0 - posf) + (14 * velf + 16 * vel0) * ts + (3 * acc0 - 2 * accf) * ts**2) / (
        2 * ts**4
    )
    a5 = (12 * (posf - pos0) - (6 * velf + 6 * vel0) * ts - (acc0 - accf) * ts**2) / (
        2 * ts**5
    )

    return a0, a1, a2, a3, a4, a5

File: analysis/cpp/src/path_generation.cpp
#include "controlmain.h"

struct QuinticCoeffs{
    double a0, a1, a2, a3, a4, a5;
};

QuinticCoeffs quintic_coeffs(double pos0, double posf, double vel0, double velf, double acc0, double accf, double ts){
    QuinticCoeffs c;
    c.a0 = pos0;
    c.a1 = vel0;
    c.a2 = acc0/2.0;
    c.a3 = (20*(posf - pos0) - (8*velf + 12*vel0)*ts - (3*acc0 - accf)*pow(ts, 2))/(2*pow(ts, 3));
    c.a4 = (30*(pos0 - posf) + (14*velf + 16*vel0)*ts + (3*acc0 - 2*accf)*pow(ts, 2))/(2*pow(ts, 4));
    c.a5 = (12*(posf - pos0) - (6*velf + 6*vel0)*ts - (acc0 - accf)*pow(ts, 2))/(2*pow(ts, 5));
    return c;
}

std::vector< std::array<double, 3> > quintic_path(double pos0, double posf, double vel0, double velf, double acc0, double accf, double ts, double h){
    QuinticCoeffs c = quintic_coeffs(pos0, posf, vel0, velf, acc0, accf, ts);
    int n_interval = std::max(1, static_cast<int>(std::round(ts/h)));
    
    std::vector< std::array<double, 3> > path;
    double pos = 0, vel = 0, acc = 0;
    for(int i = 0; i <= n_interval; i++){
        double t = ts*static_cast<double>(i)/static_cast<double>(n_interval);
        double t2 = t*t;
        double t3 = t2*t;
        double t4 = t3*t;
        double t5 = t4*t;
        pos = c.a0 + c.a1*t + c.a2*t2 + c.a3*t3 + c.a4*t4 + c.a5*t5;
        vel = c.a1 + 2 * c.a2*t + 3*c.a3*t2 + 4*c.a4*t3 + 5*c.a5*t4;
        acc = 2*c.a2 + 6*c.a3*t + 12*c.a4*t2 + 20*c.a5*t3;
        path.push_back({pos, vel, acc});
    }
    return path;
}

std::vector< std::array<double, 3> > ControlMain::path_generation(double x0, double xf, double tf, double ta, double h, bool full_quintic){
    if(full_quintic){
        std::vector< std::array<double, 3> > path = quintic_path(x0, xf, 0.0, 0.0, 0.0, 0.0, tf, h);
        return path;
    }
    else{
        double td = tf - ta;
        double vd = (xf - x0)/td;
        double xa = x0 + 0.5*ta*vd;
        double xd = xf - 0.5*ta*vd;

        std::vector< std::array<double, 3> > pv1, pv2, pv3;

        pv1 = quintic_path(x0, xa, 0.0, vd, 0.0, 0.0, ta, h);
        pv2 = quintic_path(x0, xd, vd, vd, 0.0, 0.0, td - ta, h);
        pv3 = quintic_path(xd, xf, vd, 0.0, 0.0, 0.0, tf - td, h);

        std::vector< std::array<double, 3> > path;
        path.reserve(pv1.size() + pv2.size() + pv3.size());
        path.insert(path.end(), pv1.begin(), pv1.end());
        path.insert(path.end(), pv2.begin(), pv2.end());
        path.insert(path.end(), pv3.begin(), pv3.end());

        return path;
    }
}